Code/Framework/AzCore/AzCore/Date/DateFormat.h

```diff
@@ -0,0 +1,456 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/std/base.h>
+#include <AzCore/std/chrono/chrono.h>
+#include <AzCore/std/string/fixed_string.h>
+
+namespace AZ::Date::Internal
+{
+    // Returns true if utcTime is valid duration
+    template<class Duration, class HH_MM_SS>
+    bool GetYearMonthDayWithHoursMinutesSecondsFromUTCTime(AZStd::chrono::year_month_day& yearMonthDay, HH_MM_SS& hoursMinutesSeconds,
+        const AZStd::chrono::utc_time<Duration>& utcTime)
+    {
+        // Compute year-month-day by getting days since the epoch 1970-01-01 00:00:00 UTC
+        // using the utc_clock which accounts for leap seconds
+
+        // Year-month-day is only computable from a system_clock time_point
+        // Convert the UTC time to system_clock::time_point
+        const auto utcToSysTime = AZStd::chrono::utc_clock::to_sys(utcTime);
+        const auto serialUtcDays = AZStd::chrono::time_point_cast<AZStd::chrono::days>(utcToSysTime);
+
+        const AZStd::chrono::year_month_day currentYmd{ serialUtcDays };
+        // Validate that the date is valid in the civil calendar
+        if (!currentYmd.ok())
+        {
+            return false;
+        }
+
+        // Compute a hour-minute-second(subsecond) structure from the difference in the utc time since the epoch
+        // and days since the epoch
+        // Computes the duration since the epoch representing the start of the day
+        const HH_MM_SS currentTimeOfDay{ utcToSysTime - serialUtcDays };
+        // The time of day should never be negative
+        if (currentTimeOfDay.is_negative())
+        {
+            return false;
+        }
+
+        // Update the output parameters and return success
+        yearMonthDay = currentYmd;
+        hoursMinutesSeconds = currentTimeOfDay;
+        return true;
+    }
+}
+
+namespace AZ::Date
+{
+    //! Contains methods to convert a chrono UTC clock time point into
+    //! ISO8601 timestamp per https://www.iso.org/iso-8601-date-and-time-format.html
+    using Iso8601TimestampString = AZStd::fixed_string<64>;
+
+    //! Date Formats
+    // formats an extended format ISO 8601 timestamp date
+    // Returns true if utcTime is valid
+    template<class Duration>
+    bool GetIso8601ExtendedFormatDate(Iso8601TimestampString& utcTimestamp, const AZStd::chrono::utc_time<Duration>& utcTime)
+    {
+        using DurationType = typename AZStd::remove_cvref_t<decltype(utcTime)>::duration;
+
+        AZStd::chrono::year_month_day currentYmd;
+        [[maybe_unused]] AZStd::chrono::hh_mm_ss<DurationType> currentTimeOfDay;
```

Comment:
`currentTimeOfDay` appears to have similar "usage" pattern as `utcTime`, I don't know why one would need `[[maybe_unused]]` and not the other.  We can see that `currentTimeOfDay` is an "out" parameter and `utcTime` is a read-only "in" parameter, so it does seem correct that `currentTimeOfDay` is `[[maybe_unused]]` (maybe not read from), but does the compiler investigate that `currentTimeOfDay` is only written to in the internals of `GetYearMonthDayWithHoursMinutesSecondsFromUTCTime`?  I mean, if it doesn't look into that function, how else would it know whether `hoursMinutesSeconds` is not read and written?

Reply:
The clang compiler is really smart and detects that the memory address where currentTimeOfDay resides on the stack is only being written to, but not read.
It can track this because the `GetYearMonthDayWithHoursMinutesSecondsFromUTCTime` is inlined due to be a template function.

Reply:
Okay that makes sense if that function is inlined.  Thanks!